### DISCUSSION TITLE
VGA-card for MPF-II

### DISCUSSION THREAD

**Discussion by vitasam** [markadev/AppleII-VGA#4] [General]
Hello, moving here with a [discussion](https://github.com/markadev/AppleII-VGA/pull/1) about Microprofessor-II.

I will copy your KiCAD project to my Github (with credit to the author) and start rewiring the connections for MPF-II bus.

**Comment by markadev** [comment]
That sounds good to me. Hope it goes well!

**Comment by vitasam** [comment]
Yep, KiCAD 6.0 ERC keeps failing on `Error: Input Power pin not driven by Output Powers pins`.  Is this due to [PWR_FLAG](https://electronics.stackexchange.com/questions/596978/kicad-erc-error-3-power-input-pin-not-driven-when-attached-to-transistor) missed for all output power pins? 

UPDATE: Errors have been solved by deleting the old Power Flags from `Connections` page and adding PWR_FLAG instead (from KiCAD 6.0). Remaining Warnings are mostly about "does not include the library AppleVGA-rescue"

**Reply by markadev** [reply]
Yeah, thanks for fixing that.  It's a bit odd that KiCad can't figure out that power through a diode is still power but 🤷‍♂️.  I fixed some of the other warnings about bus naming & the AppleVGA-rescue library just to tidy it up.  A few warnings still show up but they don't look like real problems so I'll just leave them alone.

**Reply by vitasam** [reply]
![20220915_175854.jpg](https://user-images.githubusercontent.com/5597505/190453232-6f4003df-292c-4772-8f84-bae4ad0d5a16.jpg)

Components collecting

**Reply by markadev** [reply]
🥳
This is my experience that you may or may not find helpful when assembling:
 - I assembled and tested the VGA output section first before tackling the bus interface.  There's a test image in the source code that you can set to display immediately at power on and it can be powered entirely from USB.
 - There are a lot of wires in the bus-to-transceiver section so I worked from shortest to longest wires and used stiff wire (CAT5 scraps) so the wires were arched up during soldering but I could route and fold them down nicely afterward and they stay in place.
 - I just used a piece of scrap CAT5 cable to connect the pin header to a DB15 connector.  If you have trouble soldering to the connector in your pic then you could also try a solder cup connector [like this](https://www.amazon.com/gp/product/B00W97OOHC/ref=ppx_yo_dt_b_asin_title_o06_s00?ie=UTF8&psc=1)

**Reply by vitasam** [reply]
Thanks for suggestions! I'm still waiting the proto-board to arrive, the rest of components are available.

**Reply by vitasam** [reply]
Hi, 
I received Apple-II prototype board, and after cutting a gap it fits MPF-II nicely.

`There's a test image in the source code that you can set to display immediately at power on and it can be powered entirely from USB.`
Could you clarify how to set the test image?

**Reply by markadev** [reply]
Sure, it does look like I forgot to add an `#ifdef` to allow it to be enabled.

There's a function named `render_testpattern` in [render.c](https://github.com/markadev/AppleII-VGA/blob/main/pico/render.c#L12-L112).  This function would need to be called by the `render_loop` function instead of the mode-rendering functions.  For example, the soft-switches should get set to 0 at boot time so you can replace the lores rendering that would have normally been used.
```
        case 0:
-           render_lores();
+           render_testpattern();
            break;
        case SOFTSW_MIX_MODE:
```

And also make sure you're building cmake's "Release" build type so that all the compiler optimizations are enabled 😀

**Reply by vitasam** [reply]
Thanks for both answers! I just applied a suggestion in my pull-request, hope it is Ok now

**Comment by vitasam** [comment]
![mpf-ii-vga_card](https://user-images.githubusercontent.com/5597505/227734592-f6bb30b5-a789-487c-b203-54c87907cb26.jpg)

The last components have arrived, and I continue to assemble the board.

**Comment by vitasam** [comment]
Hi, I continue MPF-II related discussion in this conversation.

I compiled a [comparison](https://github.com/vitasam/mpf-ii_retro_6502/blob/main/Docs/appli-ii_mpf-ii_bus_pinouts.pdf) table of Apple-II and MPF-II buses and updated the [README](https://github.com/vitasam/mpf-ii_retro_6502/blob/main/VGA-card/README.md) file.

What do you think will it be possible to re-use your VGA PCB board HW without any changes, assuming that I will make an adaptor board `MPF-II_BUS <---> Apple-II_BUS`?

(I'm thinking that for the lines that are different between MPF-II and Apple-II I will add jumpers so it is possible to disconnect the line if needed)

**Reply by markadev** [reply]
Hey there, yeah it's probably possible.  Most of the different lines aren't used by the VGA card so could be left off but there's a couple notable ones that are different:
 * pin 41 `/DEVSEL` is used by the VGA card to know when its configuration registers are being accessed.  You can leave it tied high and the card will work but you'll be unable to write any configuration changes (like changing the font/colors/etc.).  I'm not sure if the `/EXTC` signal on the MPF-II pin could be used as a replacement, or you could just hard-code some address in software to replace use of the `/DEVSEL` signal
 * pin 40 on the Apple II is the Φ0 bus clock signal, which is different but related to the 6502 Φ2 signal.  There is a diagram on page 67 of "The Apple II Circuit Description" (google for the book's PDF) that shows the relationship between them. I think you could use the "6502 Φ2" in place of the "Apple II bus Φ0" signal but you would need to adjust the timing of the PIO program to account for the difference in when the signal changes.

**Reply by vitasam** [reply]
Thank you @markadev ! 
I will start a new KiCAD project with MPF-II bus adapter, will keep updates about it in this conversation.

**Comment by vitasam** [comment]
Hi,

Here is a [Work-in-Progress](https://github.com/vitasam/mpf-ii_retro_6502/blob/main/VGA-card/MPF-II_bus_adapter/KiCAD_prj/mpf-ii-bus-adapter.pdf) of the schecmatics

Not sure that it will work but I've added a selection of Φ0 on the Apple-bus: as Φ2 or as inverted Φ1.

**Comment by vitasam** [comment]
And WIP of the PCB:
![mpf-ii-bus-adapter_pcb_rev_a](https://github.com/user-attachments/assets/11c67ac6-938d-4ef2-b261-591b24399a0c)

**Comment by vitasam** [comment]
Finally the rev.A of the adapter board arrived:

![image](https://github.com/user-attachments/assets/a2160ad3-2d51-4912-bc74-ae3fe2036b38)
